pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -507,7 +516,9 @@ func (c *Controller) preparePacket(tf *crdv1beta1.Traceflow, intf *interfacestor
 			packet.IPFlags = 0
 		}
 	} else if tf.Spec.Packet.IPHeader != nil {
-		packet.IPProto = uint8(tf.Spec.Packet.IPHeader.Protocol)
+		if tf.Spec.Destination.Service == "" {
+			packet.IPProto = uint8(tf.Spec.Packet.IPHeader.Protocol)
+		}
```

Comment:
why this? shouldn't `f.Spec.Packet.IPHeader.Protocol` take priority over the defaults?

maybe this should be:
```go
		if tf.Spec.Packet.IPHeader.Protocol > 0 {
			packet.IPProto = uint8(tf.Spec.Packet.IPHeader.Protocol)
		}
```

even though 0 is technically a valid protocol number AFAIK
It's unfortunate that we didn't make the `Protocol` field a pointer in the API.

Reply:
It is to ensure that there is no protocol mismatch by mistake. It can happen that only`tf.Spec.Packet.IPHeader.Protocol` is provided by user and it is `UDP` and the destination service supports only `TCP`, then port is default one but protocol will be user provided and it will be different than default. So, current implementation ensures that protocol and port are tightly coupled. Either both port and protocol are default one or both are provided by user.

Reply:
So if my Service has 3 ports (as an example): `ports[0] = tcp:80`, `ports[1] = udp:8080`, `ports[2] = sctp:8080`. How can I do a Traceflow for the SCTP port?

Reply:
I see that currently we don't support `SCTP` in traceflow.
You can only provide `SCTP` in `tf.Spec.Packet.IPHeader.Protocol` with no port number. It results in packet with 0 port number and `SCTP` protocol.
```
type TransportHeader struct {
	ICMP *ICMPEchoRequestHeader `json:"icmp,omitempty" yaml:"icmp,omitempty"`
	UDP  *UDPHeader             `json:"udp,omitempty" yaml:"udp,omitempty"`
	TCP  *TCPHeader             `json:"tcp,omitempty" yaml:"tcp,omitempty"`
}
```

Reply:
That makes sense. We can add SCTP in a follow-up PR.

However, I think the original point still stands. If `tf.Spec.Packet.IPHeader.Protocol > 0`, we should not just silently ignore the value. I would still craft the packet "requested" by the user.

An alternative would be some kind of error, but I think generally for the Traceflow controller we avoid errors unless it is something that prevents us from crafting the packet altogether (e.g. missing IP address).

That's not very important overall, but I see no strong reason for the special case here.

Reply:
changed to use `tf.Spec.Packet.IPHeader.Protocol` if it is > 0, thanks @antoninbas